spinn_front_end_common/interface/interface_functions/command_sender_adder.py

```diff
@@ -0,0 +1,111 @@
+# Copyright (c) 2017-2019 The University of Manchester
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from spinn_front_end_common.data.fec_data_view import FecDataView
+from spinn_front_end_common.abstract_models import (
+    AbstractSendMeMulticastCommandsVertex)
+from spinn_front_end_common.utility_models import CommandSender
+from pacman.model.graphs.application import (
+    ApplicationFPGAVertex, ApplicationSpiNNakerLinkVertex)
+from pacman.model.placements import Placement
+from pacman.model.partitioner_splitters import SplitterOneAppOneMachine
+from spinn_utilities.progress_bar import ProgressBar
+
+
+def add_command_senders(system_placements):
+    """ Add command senders
+    """
+    CommandSenderAdder(system_placements).add_command_senders()
+
+
+class CommandSenderAdder(object):
+
+    __slots__ = [
+        "__command_sender_for_chip",
+        "__general_command_sender",
+        "__system_placements"
+    ]
+
+    def __init__(self, system_placements):
+        self.__system_placements = system_placements
+
+        # Keep track of command senders by which chip they are on
+        self.__command_sender_for_chip = dict()
+        self.__general_command_sender = None
+
+    def add_command_senders(self):
+        progress = ProgressBar(FecDataView.get_n_vertices(), "Adding commands")
+        for vertex in progress.over(FecDataView.iterate_vertices()):
+            if isinstance(vertex, AbstractSendMeMulticastCommandsVertex):
+                machine = FecDataView.get_machine()
+                link_data = None
+
+                # See if we need a specific placement for a device
+                if isinstance(vertex, ApplicationFPGAVertex):
```

Comment:
Consider moving class specific code to the classes

Reply:
This has been moved now